include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h

```diff
@@ -0,0 +1,243 @@
+//===--- UpdatingInstructionIterator.h --------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// Classes for tracking instruction iterators to be updated when instructions
+/// are added or deleted.
+///
+/// UpdatingInstructionIteratorRegistry typically resides in the currently
+/// active InstructionDeleter object.
+///
+/// UpdatingInstructionIterator is the iterator adapter.
+/// It is produced by: UpdatingInstructionIteratorRegistry::makeIterator()
+///
+/// Iterators are typically encapsulated in a range returned by
+/// UpdatingInstructionIteratorRegistry::iteratorRange() for use in
+/// range-based for loops:
+///
+///   for (SILInstruction *inst : registry.iteratorRange(bb)) ...
+///
+/// Or more commonly, directly from the InstructionDeleter:
+///
+///   for (SILInstruction *inst : deleter.iteratorRange(bb)) ...
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SILOPTIMIZER_UTILS_UPDATINGINSTRUCTIONITERATOR_H
+#define SWIFT_SILOPTIMIZER_UTILS_UPDATINGINSTRUCTIONITERATOR_H
+
+#include "swift/SIL/SILInstruction.h"
+#include <utility>
+
+namespace swift {
+
+class UpdatingInstructionIteratorRegistry;
+
+namespace detail {
+
+struct UpdatingInstructionIteratorState {
```

Comment:
Is it really worth separating the state and the iterator into two separate classes and doing reference counting? It adds quite a bit of complexity. Can't you just add references to all iterators into the notification-receiving list?

Also, I'm not sure if this really works. After this code
```
  iterator2 = iterator1
  iterator1++
``` 
both iterators still share the same state, but should refer to different instructions

Reply:
Sorry, what I pushed last night was an incomplete mish-mash. I originally had a next-style iterator as an RAII object that registered and deregistered itself. I played with adding an iterator adapter to see if the range-based API would work, but haven't finished. My biggest concern now is that an iterator is inadvertently copied on each iteration. I was considering prohibiting post-increment and keeping the registry efficient. The alternative is dropping the iterator ID and registering a pointer to each iterator. I'm concerned about the overhead, but I'll try that and see how bad it is since there won't be any weird pitfalls with that approach.

Reply:
@eeckstein ok, here's the finished UpdatingIterator template. I tested the reverse iterators locally. Everything has been converted over to normal iterator values now. No reference counting, no separate state. We just need to be careful not to create lots of instruction iterator copies on the critical path. So far it isn't a problem.

https://github.com/apple/swift/pull/37669/commits/c05416a17ec602839de97dab88e1b01f890d8221